src/Microsoft.Health.Fhir.SqlServer/IndentedStringBuilder.cs

```diff
@@ -0,0 +1,65 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.CodeDom.Compiler;
+using System.Text;
+using EnsureThat;
+
+namespace Microsoft.Health.Fhir.SqlServer
+{
+    /// <summary>
+    /// A wrapper around <see cref="StringBuilder"/> that provides methods that insert a indent string and track the current indentation level.
+    /// Inspired by <see cref="IndentedTextWriter"/>
+    /// </summary>
+    internal partial class IndentedStringBuilder
+    {
+        private const string IndentString = "    ";
+
+        private bool _indentPending;
+        private int _indentLevel;
+
+        internal int IndentLevel
+        {
+            get => _indentLevel;
+            set => _indentLevel = EnsureArg.IsGte(value, 0);
+        }
+
+        /// <summary>
+        /// Increments the indent level until the returned object is disposed.
+        /// </summary>
+        /// <returns>A scope to be disposed when the indentation level is to be restored</returns>
+        internal IndentedScope Indent() => new IndentedScope(this);
+
+        protected virtual void AppendIndent()
+        {
+            if (_indentPending)
+            {
+                for (int i = 0; i < _indentLevel; i++)
+                {
+                    _inner.Append(IndentString);
```

Comment:
Could do something like `_inner.Append(new String(' ', _indentLevel * 4);`

Reply:
That allocates a new string every time, so will be less efficient

Reply:
Or could pad the number of spaces.

Reply:
Same issue I suppose.